compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -3360,7 +3361,40 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                 }
 
                 match arg.layout.abi {
-                    Abi::Aggregate { .. } => {}
+                    Abi::Aggregate { .. } => {
+                        // Pass and return structures up to 2 pointers in size by value,
+                        // matching `ScalarPair`. LLVM will usually pass these in 2 registers
+                        // which is more efficient than by-ref.
+                        let ptr_size = Pointer.size(self);
+                        let max_by_val_size = ptr_size * 2;
+                        let size = arg.layout.size;
+
+                        if arg.layout.is_unsized() || size > max_by_val_size {
```

Comment:
The old code used `size > ptr_size`. You changed it to `size > ptr_size * 2`.

Reply:
Oops, you're right I forgot about this. It explain the performance regressions.
I will revert this change. Thanks for finding this out.